src/storage/client/StorageClient.inl

```diff
@@ -151,6 +151,84 @@ folly::SemiFuture<StorageRpcResponse<Response>> StorageClient::collectResponse(
             });
         });  // via
     }  // for
+    if (UNLIKELY(context->finishSending())) {
+        // Received all responses, most likely, all rpc failed
+        context->promise.setValue(std::move(context->resp));
+    }
+
+    return context->promise.getSemiFuture();
+}
+
+
+// \TODO(shylock) more general?
+template<class Request, class RemoteFunc, class Response>
+folly::SemiFuture<StorageRpcResponse<std::pair<HostAddr, Response>>>
+StorageClient::collectResponseWithoutLeader(
+        folly::EventBase* evb,
+        std::unordered_map<HostAddr, Request> requests,
+        RemoteFunc&& remoteFunc) {
+    auto context = std::make_shared<ResponseContext<
+            Request, RemoteFunc, std::pair<HostAddr, Response>>>(
+        requests.size(), std::move(remoteFunc));
+
+    if (evb == nullptr) {
+        DCHECK(!!ioThreadPool_);
+        evb = ioThreadPool_->getEventBase();
+    }
+
+    for (auto& req : requests) {
+        auto& host = req.first;
+        auto res = context->insertRequest(host, std::move(req.second));
+        DCHECK(res.second) << "Empty RPC request!";
+        // Invoke the remote method
+        folly::via(evb, [this, evb, context, host, /*spaceId,*/ res] () mutable {
+            auto client = this->clientsMan_->client(host, evb);
+            // Result is a pair of <Request&, bool>
+            context->serverMethod(client.get(), *res.first)
+            // Future process code will be executed on the IO thread
+            // Since all requests are sent using the same eventbase, all then-callback
+            // will be executed on the same IO thread
+            .then(evb, [/*this,*/ context, host/*, spaceId*/] (folly::Try<Response>&& val) {
+                if (UNLIKELY(val.hasException())) {
+                    LOG(ERROR) << "Request to " << host << " failed: " << val.exception().what();
+                    context->resp.markFailure();
+                } else {
+                    auto resp = std::move(val.value());
+                    auto& result = resp.get_result();
+                    bool hasFailure{false};
+                    for (auto& code : result.get_failed_codes()) {
+                        VLOG(3) << "Failure! Failed part " << code.get_part_id()
+                                << ", failed code " << static_cast<int32_t>(code.get_code());
+                        hasFailure = true;
+                        if (UNLIKELY(code.get_code() ==
+                                storage::cpp2::ErrorCode::E_LEADER_CHANGED)) {
+                            DCHECK(false) << "Impossible leader require!";
```

Comment:
what mean about DCHECK(false)?

Reply:
It should be a impossible case, so I add dcheck for checking.